showpath.1
.Dd December 3, 2012
.Dt SHOWPATH 1 LOCAL
.Os Generic Unix
.
.Sh NAME
.Nm showpath
.Nd manageably edit path
.
.Sh SYNOPSIS
.Nm showpath
.Op Fl t Ar type | Fl e Ar ENV_NAME
.Op Fl s Ar separator
.Ar path_entry ...
.
.Sh DESCRIPTION
The showpath tool writes a string suitable for use as a path
to stdout.
The string is constructed by combining the path entries given on
the command line, in order, with duplicates pruned.
.
.Pp
A path_entry containing the path separator is treated as a fragment
of a path and is broken down into its constituent parts by splitting
at the separators; these parts will then be added as if they were
named separately on the command line.
.
.Pp
A path_entry beginning with
.Ql \&%
is a
.Em short name ,
and will be expanded to a set of path entries that are then added
as if they were named separately on the command line.  The short
names currently supported are
.Ql %current ,
which expands to the entries in the current path (taken from the
environment), and
.Ql %pwd ,
which expands to the current directory.
Allowing additional short names to be defined (by system and/or
user configuration files) is on the to-do list.
.
.Pp
If a path entry (including one that is the result of short name
expansion or path splitting) begins with
.Ql \&~ ,
the
.Ql \&~
will be expanded to the current value of
.Ev $HOME .
.
.Pp
A path_entry prefixed with
.Ql \&^
will have shortname expansion or path splitting done as normal,
and then the set of path entries it expands to will be
.Em removed
from the path being constructed instead of being added to it.
Entries that have been removed may be re-added, and will be placed
at the end of the path (this is exactly the same handling as anything
else not in the path under construction).
.
.Pp
The following options are available:
.Bl -tag -width flag
.
.It Fl e Ar ENV_NAME
Specifies an environment variable name to look up for the expansion of
.Ql %current .
.
.It Fl s Ar separator
Specifies the path separator.
The separator must be a single character and defaults to
.Ql \&: .
.
.It Fl t Ar type
Specifies a pre-configured path type.
Valid types are currently compiled in; making them configurable (by system
and/or user configuration) is on the to-do list.
.Pp
Currently known types are:
.Bl -tag -compact
.It Li exec
(for
.Ev $PATH )
.It Li man
(for
.Ev $MANPATH )
.El
.
.Pp
The
.Fl t
option currently only provides a shorthand for an equivalent use of the
.Fl e
option.
Adding short names for sets of path entries is on the to-do list;
when this happens, specifying a type will also select a set of short
names based on the type.
.
.El
.Pp
The
.Fl t
and
.Fl e
options are mutually exclusive, and at most one of either can be specified.
Specifying neither is equivalent to specifying
.Bk -words
.\" When being processed with troff, this Ql expands to a call to Li, which
.\" generates a "cancels effect" warning since its first arg is a macro.
.\" But Ql does its thing on the the '-t exec' as desired (with nroff it
.\" quotes it), which is why it's there.   --DV
.Ql Fl t Li exec .
.Ek
.
.Sh FILES
None yet.
.
.Sh EXAMPLES
Put
.Pa ~/bin
at the beginning of
.Ev $PATH ,
or move it to the beginning if it's already in there somewhere:
.Bd -literal -offset indent -compact
PATH="\*(gashowpath ~/bin %current\*(ga"
.Ed
.
.Pp
Put
.Pa ~/bin
at the end of
.Ev $PATH ,
or move it to the end if it's already in there somewhere:
.Bd -literal -offset indent -compact
PATH="\*(gashowpath %current ^~/bin ~/bin\*(ga"
.Ed
.
.Pp
Tell gcc to use header and library files from Macports,
without clobbering existing values of the environment
variables used:
.Bd -literal -offset indent -compact
C_INCLUDE_PATH="\*(gashowpath -eC_INCLUDE_PATH %current /opt/local/include\*(ga"
LIBRARY_PATH="\*(gashowpath -eLIBRARY_PATH %current /opt/local/lib\*(ga"
export C_INCLUDE_PATH LIBRARY_PATH
.Ed
.
.Pp
Invoke a setup script that adds entries to
.Ev $PATH ,
but force new entries to go at the end instead of the beginning:
.Bd -literal -offset indent -compact
oldpath="$PATH"
\&. /path/to/setup/script
PATH="\*(gashowpath "$oldpath" %current\*(ga"
.Ed
.
.Sh HISTORY
The MFCF environment at the University of Waterloo has a tool by
the same name with similar behavior, which simplifies configuring
paths for the local environment.
.
.Pp
This version does
.Em not
handle its arguments in a manner compatible with the UW version.
.
.Sh AUTHORS
.An "Dave Vandervies" Aq dj3vande@eskimo.com
.
.Pp
.An -nosplit
With contributions from
.An "Stephen Paul Weber" Aq singpolyma@singpolyma.net
.
.Sh BUGS
Empty entries are discarded by strtok when breaking down a path
fragment into its constituent parts.
If you really want an empty path entry, give yourself a whack with
the cluestick and then say
.Bk -words
.Ql showpath %current \&"" .
.Ek
.
.Pp
The
.Li ~ Ns -expansion
done by
.Nm
prevents any path entry in the output from starting with
.Ql \&~ .
.Pp
Send new bug reports to
.Aq dj3vande@eskimo.com .


showpath.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <unistd.h>	/*for getopt and getcwd*/

#include "showpath.h"

/* stores argv[0] for use in output */
static char *myname;

static struct config
{
	char *envname;
	char pathsep;
} config;

static void set_defaults(void)
{
	config.envname="PATH";
	config.pathsep=':';
}


/*Expand shortnames (path_entries beginning with '%')
  Currently handles '%current' (existing entry from environment) and
    '%pwd' (working directory at time of showpath invocation).
  When we add config file capability, this will be where we look up
    the path entries for shortnames.
*/
struct path_list *expand_shortname(const char *name)
{

	if(strcmp(name,"%current")==0)
	{
		const char *env_string=getenv(config.envname);
		if(env_string)
			return split_path(env_string,config.pathsep);
		else
			return alloc_entry();	/*no path string is not error*/
	}
	else if(strcmp(name,"%pwd")==0)
	{
		char *pwd;
		struct path_list *ret;

		pwd=getcwd(NULL,-1);
		if(!pwd)
			return NULL;
		ret=alloc_entry();
		if(!ret)
			return NULL;

		if(insert_entry(ret,pwd))
		{
			free_entry(ret);
			ret=NULL;
		}

		free(pwd);
		return ret;
	}
	else
	{
		fprintf(stderr,"%s: Unrecognized shortname '%s'\n",myname,name);
		return NULL;
	}
}


/*Set the type of path we're building.
  Used for -t switch.
  Currently only affects config.envname.
*/
int set_type(const char *type)
{
	const char *types[]={"exec","man",NULL};
	char *envs[]={"PATH","MANPATH",NULL};
	size_t i = -1;
	while(types[++i])
	{
		if(strcmp(types[i],type)==0)
		{
			config.envname=envs[i];
			return 0;
		}
	}

	return 1;
}

void shortusage(void)
{
	printf("Usage: %s [-t type | -e env_name] [-s separator ] path-entry ...\n",myname);
}

void usage(void)
{
	shortusage();
	printf("  Outputs a path containing all path-entries on the command line, in\n");
	printf("    the order given, with duplicates removed.\n");
	printf("  Known types are 'exec' (for $PATH) and 'man' (for $MANPATH).\n");
	printf("  The shortname '%%current' expands to the current value from\n");
	printf("    the environment.\n");
	printf("  The shortname '%%pwd' expands to the working directory at the time\n");
	printf("    of invocation of %s.\n",myname);
}


/*Add a path entry, handling shortname expansion and path splitting
    as appropriate
*/
int add_entry(struct path_list *pl,const char *ent)
{
	struct path_list *new;
	int ret;
	int remove=0;

	if(ent[0]=='^')
	{
		remove=1;
		ent++;
	}

	if(ent[0]=='%')
		new=expand_shortname(ent);
	else if(strchr(ent,config.pathsep))
		new=split_path(ent,config.pathsep);
	else	/*simple entry, just add*/
	{
		if(remove)
		{
			remove_entry(pl,ent);
			return 0;
		}
		else
		{
			return insert_entry(pl,ent);
		}
	}

	/*If we get here, we have a list of entries to add in new
	    (or NULL if we've failed somewhere along the line)
	*/
	if(!new)
		return -1;

	if(remove)
	{
		remove_entries(pl,new);
		ret=0;
	}
	else
	{
		ret=insert_entries(pl,new);
	}
	free_entry(new);
	return ret;
}

int main(int argc,char **argv)
{
	int i;
	int opt;
	int have_type=0;
	struct path_list *pl;

	myname=argv[0];

	set_defaults();

	while((opt=getopt(argc,argv,":t:e:s:"))!=-1)
	{
		switch(opt)
		{
		case 't':
			if(have_type)
			{
				fprintf(stderr,"%s: Only one '-t' or '-e' allowed!\n",myname);
				shortusage();
				return EXIT_FAILURE;
			}
			if(set_type(optarg))
			{
				fprintf(stderr,"%s: Unknown type '%s'!\n",myname,optarg);
				shortusage();
				return EXIT_FAILURE;
			}
			have_type=1;
			break;
		case 'e':
			if(have_type)
			{
				fprintf(stderr,"%s: Only one '-t' or '-e' allowed!\n",myname);
				shortusage();
				return EXIT_FAILURE;
			}
			config.envname=optarg;
			have_type=1;
			break;
		case 's':
			if(optarg[1])
			{
				fprintf(stderr,"%s: Separator must be a single character!\n",myname);
				shortusage();
				return EXIT_FAILURE;
			}
			config.pathsep = optarg[0];
			break;
		case ':':
			fprintf(stderr,"%s: Option '%c' requires an argument\n",myname,optopt);
			shortusage();
			return EXIT_FAILURE;
		case '?':
			fprintf(stderr,"%s: Unrecognized option: %c\n",myname,optopt);
			shortusage();
			return EXIT_FAILURE;
		default:
			usage();
			return EXIT_FAILURE;
		}
	}

	/* These lines allow us to treat argc and argv as though 
	 * any switches were not there
	 */
	argc-=optind;
	argv+=optind;

	if(argc<1)
	{
		usage();
		return 0;
	}

	pl=alloc_entry();
	if(!pl)
	{
		perror("alloc_entry");
		exit(EXIT_FAILURE);
	}

	for(i=0;i<argc;i++)
	{
		int ret=add_entry(pl,argv[i]);
		if(ret)
		{
			/*XXX This will get ugly if we have a long path
			    on the command line.  Is there a better way
			    to handle that?   --DV
			*/
			fprintf(stderr,"%s: Adding path entry '%s' failed!\n",myname,argv[i]);
			exit(EXIT_FAILURE);
		}
	}

	if(pl->num)
	{
		size_t idx;
		printf("%s",pl->entries[0]);
		for(idx=1;idx<pl->num;idx++)
			printf("%c%s",config.pathsep,pl->entries[idx]);
		printf("\n");
	}

	free_entry(pl);

	return 0;
}
